### PR TITLE
Cross provider model comparisons: GPT4 vs Llama2

### PR DESCRIPTION
**Pull request by steventkrawczyk**
Important files:
* examples/notebooks/GPT4vsLlama2.ipynb
* prompttools/experiment/experiments/llama_cpp_experiment.py
* prompttools/experiment/experiments/openai_chat_experiment.py
* prompttools/selector/prompt_selector.py

### CODE REVIEW

File: prompttools/experiment/experiments/experiment.py [hunk 1]
```diff
@@ -124,7 +124,8 @@ def run(
             for _ in range(runs):
                 self.queue.enqueue(
                     self.completion_fn,
-                    combo,
+                    # We need to filter out defaults that are invalid JSON from the request
+                    {k: v for k, v in combo.items() if v and v != float("inf")},
```
Comment (NivekT) [inline prompttools/experiment/experiments/experiment.py:128]: 1. What the common invalid values for `v`? Where do they come from?
2. Is it possible to have `v = 0` or `v = None` that are valid?
3. If I understand correctly, this merely throws away the `k: v` pair but keeps the rest of the combo?

File: examples/notebooks/GPT4vsLlama2.ipynb [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    "Draft a short sales email, 50 words or less, asking a prospect for 15 minutes\n",
+    "of their time to chat about how they're using large language models.\n",
+    "\"\"\",\n",
+    "\"\"\"\n",
+    "Do you offer refunds?\n",
+    "\"\"\",\n",
+    "\"\"\"\n",
+    "Is 7 a prime number?\n",
+    "\"\"\"]\n",
+    "selectors = [PromptSelector(instructions[i], inputs[i]) for i in range(3)]"
```
Comment (NivekT) [inline examples/notebooks/GPT4vsLlama2.ipynb:151]:
```suggestion
    "selectors = [PromptSelector(instructions[i], inputs[i]) for i in range(len(inputs))]"
```

File: examples/notebooks/GPT4vsLlama2.ipynb [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    "openai_experiment = OpenAIChatExperiment(models, selectors, temperature=temperatures)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "6c3162e6",
+   "metadata": {},
+   "source": [
+    "After that - we define our harness to run experiments"
```
Comment (NivekT) [inline examples/notebooks/GPT4vsLlama2.ipynb:194]: Let's add a short explanation about what `MultiExperimentHarness` does

**Review by NivekT — APPROVED** [review]
Thanks! Just a few comments. Have a look at the filter in `experiment.py`

File: prompttools/experiment/experiments/experiment.py (hunk 1 above)
Comment (steventkrawczyk) [reply prompttools/experiment/experiments/experiment.py:128]: We have some default values that work in python but can't be serialized into JSON and sent over the wire (e.g. `None` or `float("inf")`. We need to toss these out before making the request. The result will be that OpenAI uses the corresponding default for the missing value, so the request should be equivalent to what we track on our end.

Comment (steventkrawczyk) [reply prompttools/experiment/experiments/experiment.py:128]: We should account for `v=0` though, we can keep those.

And yeah, we'll toss any pair that's `None` or `float("inf")` and keep the rest.

**Pull request merged**